source/geometry_model/spherical_chunk.cc

```diff
@@ -0,0 +1,416 @@
+          break;
+        }
+        default:
+          Assert (false, ExcNotImplemented ());
+      }
+    return output_vertex;
+  }
+
+  template <int dim>
+  Point<dim>
+  SphericalChunkGeometry<dim>::
+  pull_back(const Point<dim> &v) const
+  {
+    Point<dim> output_vertex;
+    switch (dim)
+      {
+        case 2:
+        {
+          output_vertex[0] = std::atan2(v[0], v[1]);
+          output_vertex[1] = std::sqrt(v[0]*v[0] + v[1]*v[1]);
+          break;
+        }
+        case 3:
+        {
+          double radius=std::sqrt(v[0]*v[0] + v[1]*v[1] + v[2]*v[2]);
+          output_vertex[0] = std::atan2(v[0], v[2]);
+          output_vertex[1] = std::asin(v[1]/radius);
+          output_vertex[2] = radius;
+          break;
+        }
+        default:
+          Assert (false, ExcNotImplemented ());
+      }
+    return output_vertex;
+  }
+
+}
+
+
+namespace aspect
+{
+  namespace GeometryModel
+  {
+
+    template <int dim>
+    void
+    SphericalChunk<dim>::
+    create_coarse_mesh (parallel::distributed::Triangulation<dim> &coarse_grid) const
+    {
+      std::vector<unsigned int> rep_vec(repetitions, repetitions+dim);
+      GridGenerator::subdivided_hyper_rectangle (coarse_grid,
+                                                 rep_vec,
+                                                 point1,
+                                                 point2,
+                                                 true);
+
+      //Tell p4est about the periodicity of the mesh.
+      std::vector<GridTools::PeriodicFacePair<typename parallel::distributed::Triangulation<dim>::cell_iterator> >
+      periodicity_vector;
+      if (periodic)
+        GridTools::collect_periodic_faces
+        ( coarse_grid, /*b_id1*/ 0, /*b_id2*/ 1,
+          /*direction*/ 0, periodicity_vector);
+
+      if (periodicity_vector.size() > 0)
+        coarse_grid.add_periodicity (periodicity_vector);
+
+
+      // Transform box into spherical chunk
+      GridTools::transform (std_cxx11::bind(&SphericalChunkGeometry<dim>::push_forward,
+                                            std_cxx11::cref(manifold),
+                                            std_cxx11::_1),
+                            coarse_grid);
+
+      // Deal with a curved mesh
+      // Loop over cells,faces and set boundary indicators
+      for (typename Triangulation<dim>::active_cell_iterator cell = coarse_grid.begin_active();
+           cell != coarse_grid.end(); ++cell)
+        for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
+          if (cell->face(f)->at_boundary())
+            cell->face(f)->set_all_boundary_indicators(f);
+
+      // Attach the real manifold to slot 15. we won't use it
+      // during regular operation, but we set manifold_ids for all
+      // cells, faces and edges immediately before refinement and
+      // clear it again afterwards
+      coarse_grid.set_manifold (15, manifold);
```

Comment:
This is copied from the spherical shell. We need the manifold for mesh refinement, but manifolds do not provide normal vectors for the Q4 mapping, so we remove the manifold and fall back to the StraightBoundary object during regular operation.